src/Microsoft.OData.Core/Json/JsonLightInstanceAnnotationWriter.cs

```diff
@@ -84,27 +84,97 @@ internal JsonLightInstanceAnnotationWriter(ODataJsonLightValueSerializer valueSe
         /// <param name="ignoreFilter">Whether to ignore the filter in settings.</param>
         /// <param name="propertyName">The name of the property this instance annotation applies to</param>
         internal void WriteInstanceAnnotations(
-            IEnumerable<ODataInstanceAnnotation> instanceAnnotations,
+            ICollection<ODataInstanceAnnotation> instanceAnnotations,
             InstanceAnnotationWriteTracker tracker,
             bool ignoreFilter = false,
             string propertyName = null)
         {
             Debug.Assert(instanceAnnotations != null, "instanceAnnotations should not be null if we called this");
             Debug.Assert(tracker != null, "tracker should not be null if we called this");
 
-            HashSet<string> instanceAnnotationNames = new HashSet<string>(StringComparer.Ordinal);
-            foreach (var annotation in instanceAnnotations)
+            // this method runs in a hot path hence the optimizations
+
+            // initialize to null to avoid instantiating a HashSet if there are no annotations
+            HashSet<string> instanceAnnotationNames;
+            if (instanceAnnotations?.Count > 0)
+            {
+                instanceAnnotationNames = new HashSet<string>(StringComparer.Ordinal);
+            }
+            else
+            {
+                return;
+            }
+
+            // this method is called with a List most of the time
+            // foreach against a List does not allocate the enumerator to the heap,
+            // but foreach against an IEnumerable does
+            if (instanceAnnotations is List<ODataInstanceAnnotation> instanceAnnotationsList)

```

Comment:
Also since you set the param to be an `ICollection` means someone can no longer pass an `IEnumerable` - which again makes the if/else check/cast unnecessary.

Having said that, I imagine that any previous call to this that could have been passing in an `IEnumerable` would be failing the tests

Reply:
Doing the foreach against `ICollection` would still cause enumerator allocations. Furthermore, `ICollection` does not exposer an indexer, which would have allowed us to use a simple `for (int i...)` loop. Yes, now we can not pass `IEnumerable` to this method. But the method is only used internally, and in all cases, the argument passed in is an `ICollection`.